### PR TITLE
Use icons of the most recently used controller

### PR DESCRIPTION
**Pull request by jrb0001**
The current logic always uses the icons from the first controller. This PR changes the behavior to use the most recently used/connected controller. If it is no longer connected, then it falls back to the first controller (old logic).

When mapping actions to buttons, it uses the mapping for that specific controller. If there is no such mapping, it falls back to the first one it finds (old logic).

This helps in various situations:
- Multiple controllers of different types are connected at the same time (currently uses icons for the first controller).
- A new controller is connected before the old controller is disconnected (currently uses fallback icons).
- Godot incorrectly detects some hardware as joypad and it ends up before the first "real" controller (for example wacom drawing tabled+pen as joypad 0+1, results in fallback icons).
- An action is mapped differently on different controllers (currently takes the button for the first action it finds). For example if a game wants to swap A/B only on Nintendo controllers.

It should also work correctly when something like #58 is implemented. I wasn't sure about the best way to expose it and don't need it myself so I didn't implement it yet.

### CODE REVIEW

File: addons/controller_icons/ControllerIcons.gd [hunk 1]
```diff
@@ -71,9 +72,10 @@ var _builtin_keys := [
 	"input/ui_up",
 ]
 
-func _set_last_input_type(__last_input_type):
+func _set_last_input_type(__last_input_type, __last_controller):
 	_last_input_type = __last_input_type
-	emit_signal("input_type_changed", _last_input_type)
+	_last_controller = __last_controller
+	emit_signal("input_type_changed", _last_input_type, __last_controller)
```
Comment (rsubtil) [inline addons/controller_icons/ControllerIcons.gd:78]: For consistency with the previous parameter, use the instance variable here:
```suggestion
	emit_signal("input_type_changed", _last_input_type, _last_controller)
```

File: addons/controller_icons/ControllerIcons.gd [hunk 2]
```diff
@@ -164,12 +171,15 @@ func _add_custom_input_action(input_action: String, data: Dictionary):
 
 func refresh():
 	# All it takes is to signal icons to refresh paths
-	emit_signal("input_type_changed", _last_input_type)
+	emit_signal("input_type_changed", _last_input_type, _last_controller)
+
+func get_joypad_type(controller: int = _last_controller) -> ControllerSettings.Devices:
```
Comment (rsubtil) [inline addons/controller_icons/ControllerIcons.gd:176]: Is is necessary to expose this? Can you share what's your use-case for this?

File: addons/controller_icons/ControllerIcons.gd [hunk 3]
```diff
@@ -240,24 +250,29 @@ func get_path_type(path: String) -> PathType:
...
 			"InputEventJoypadButton", "InputEventJoypadMotion":
 				if input_type == InputType.CONTROLLER:
-					return event
-	return null
+					if event.device == controller:
+						return event
+					elif fallback == null:
+						fallback = event
+
+	return fallback
```
Comment (rsubtil) [inline addons/controller_icons/ControllerIcons.gd:273]: Functionally, this has the same behavior from the previous code; if the `if` is `true`, return `event`; else, set `fallback` to `event`, and return `fallback`, which is `event`. Why was this needed?

File: addons/controller_icons/Mapper.gd [hunk 4]
```diff
@@ -32,8 +32,16 @@ func _convert_joypad_path(path: String, fallback) -> String:
...
-func _get_joypad_type(fallback):
-	var controller_name = Input.get_joy_name(0)
+func _get_joypad_type(device, fallback):
+	var available = Input.get_connected_joypads()
+	if available.is_empty():
+		return fallback;
+	if !available.has(device):
+		device = ControllerIcons._last_controller
+	if !available.has(device):
+		device = available.front();
```
Comment (rsubtil) [inline addons/controller_icons/Mapper.gd:42]: - Add some comments here so it's clearer what these if tests are doing
- More idiomatic way to test element presence: `if not device in available:`
- Delete extra `;`
```suggestion
	# If the requested joypad is not on the connected joypad list, try using the last known connected joypad
	if not device in available:
		device = ControllerIcons._last_controller
	# If that fails too, then use whatever joypad we have connected right now
	if not device in available:
		device = available.front()
```

File: addons/controller_icons/Mapper.gd [hunk 5]
```diff
@@ -32,8 +32,16 @@ func _convert_joypad_path(path: String, fallback) -> String:
 		_:
 			return ""
 
-func _get_joypad_type(fallback):
-	var controller_name = Input.get_joy_name(0)
+func _get_joypad_type(device, fallback):
+	var available = Input.get_connected_joypads()
+	if available.is_empty():
+		return fallback;
```
Comment (rsubtil) [inline addons/controller_icons/Mapper.gd:38]: Delete extra `;`
```suggestion
		return fallback
```

**Review by rsubtil — CHANGES_REQUESTED** [review]
Thank you for your contribution! The changes look good, and while I cannot test this myself (I only have one controller), I have reviewed the code and everything looks correct, with only some small changes to be done.

I can have a look after this PR at implementing #58, since it should be quite easy to do after it.

File: addons/controller_icons/ControllerIcons.gd (hunk 2 above)
Comment (jrb0001) [reply addons/controller_icons/ControllerIcons.gd:176]: Right, it isn't related to this PR but I forgot to remove it. My usecase is to detect switch controllers to give them a different button mapping.

Please let me know if you prefer to remove it, keep it or get a separate PR.

File: addons/controller_icons/ControllerIcons.gd (hunk 3 above)
Comment (jrb0001) [reply addons/controller_icons/ControllerIcons.gd:273]: It is the same behavior only if all mappings use the "all devices" option. In case of device specific mappings, this makes it pick the correct mapping for the passed/most recently used controller. This is useful for stuff like local multiplayer or giving some controller types a different layout (like the mentioned A/B swap for Nintendo).

I was thinking about adding a `event.device == -1` check so it can only fall back to "all devices" mappings. I decided to use the current behavior as fallback, but adding that might actually be better (especially in the local multiplayer scenario with per player/controller mappings: for example some fighting games offer an optional shortcut instead of pressing multiple buttons simultaneously). What do you think?

File: addons/controller_icons/Mapper.gd (hunk 4 above)
Comment (jrb0001) [reply addons/controller_icons/Mapper.gd]: Oh nice, I didn't know that `in` can be used this way in gdscript.

File: addons/controller_icons/ControllerIcons.gd (hunk 2 above)
Comment (rsubtil) [reply addons/controller_icons/ControllerIcons.gd:176]: I assumed there was a valid reason for this, just wanted a confirmation that this was intended and not leftovers.

Your use case is valid, and not having this would make for some nasty code that requires access to private fields/methods, so I think it's good to stay :+1:

File: addons/controller_icons/ControllerIcons.gd (hunk 3 above)
Comment (rsubtil) [reply addons/controller_icons/ControllerIcons.gd]: Ah yeah, the behavior is different since this is running on a for loop. I didn't notice it before, my bad.

> I was thinking about adding a event.device == -1 check so it can only fall back to "all devices" mappings. 

Do you mean adding this as an additional check, or replacing the current one? Either way, I trust your judgement whether to add it or not; since it seems to be relevant for the future when implementing #58, it will be easier to test its behavior once that is implemented.

Comment (jrb0001) [reply addons/controller_icons/ControllerIcons.gd]: I added it as an additional check so it takes the first fallback mapping, just like it does for the device specific mappings.

Showing all matching mappings would be a cool feature but it might take a few weeks until I find time to implement it (needs a bigger refactoring of my game).

**Comment by rsubtil** [comment]
Thank you so much for your contribution! :heart:
